recipes/dev/generate_v2.py

```diff
@@ -0,0 +1,193 @@
+                "To use batch_to_device, all elements in the batch must be a dict or Tensor"
+            )
+
+
+class InferenceRecipe:
+    """
+    Recipe for generating tokens from a dense Transformer-based LLM.
+    This works for text-only generation and image-text generation.
+
+    Warning:
+        This has only been extensively tested with the following configs:
+            - multimodal_generation
+
+    This *does not* currently support the following features:
+        - torch.compile
+        - quantization through torchao
+        - speculative decoding
+        - multi-GPU generation
+    """
+
+    def __init__(self, cfg: DictConfig) -> None:
+        self._device = utils.get_device(device=cfg.device)
+        self._dtype = training.get_dtype(dtype=cfg.dtype, device=self._device)
+        self._logger = utils.get_logger(cfg.log_level)
+        training.set_seed(seed=cfg.seed)
+
+    def setup(self, cfg: DictConfig) -> None:
+        _checkpointer = config.instantiate(cfg.checkpointer)
+        _ckpt_dict = _checkpointer.load_checkpoint()
+
+        with training.set_default_dtype(self._dtype), self._device:
+            model = config.instantiate(model_cfg)
+        model.load_state_dict(_ckpt_dict[MODEL_KEY])
+
+        # Validate model was loaded in with the expected dtype.
+        training.validate_expected_param_dtype(
+            model.named_parameters(), dtype=self._dtype
+        )
+        self._logger.info(f"Model was initialized with precision {self._dtype}.")
+
+        # Ensure the cache is setup on the right device
+        with self._device:
+            model.setup_caches(batch_size=1, dtype=self._dtype)
+
+        self.model_transform = config.instantiate(cfg.transform)
+        self.to_messages = SingleTurnYAMLToMessages()
+
+    @torch.inference_mode()
+    def generate(self, cfg: DictConfig):
+        # 1. Convert input to messages
+        messages = self.to_messages(cfg.prompt)
+        is_multimodal_input = any([m.contains_media for m in messages])
+
+        # 2. Apply model transform
+        model_inputs = self.model_transform({"messages": messages})
+
+        # 3. Collate to batch size of 1 and tensor-ify
+        if is_multimodal_input:
+            batch = padded_collate_tiled_images_and_mask(
+                [model_inputs], pad_direction="left"
+            )
+        else:
+            batch = {
+                "tokens": left_pad_sequence(
+                    [torch.tensor(model_inputs["tokens"])],
+                    batch_first=True,
+                )
+            }
+        batch_to_device(batch, self._device)
+
+        # 4. Prefill step
+        generated_tokens = []
+        t0 = time.perf_counter()
+        logits = self.model(**batch)[:, -1]
+        token = sample(logits, cfg.temperature, cfg.top_k)
+        generated_tokens.append(token.item())
+
+        if is_multimodal_input:
+            cache_mask = {"encoder_mask": batch["encoder_mask"][:, -1:]}
```

Comment:
shape comments here would be nice

Reply:
Also dumb q: how should I think about `batch["encoder_mask"][:, -1:]`? Like is this just some kind of stand-in for "here are all the text tokens seen by the final image"? And why do we not need to update it as we generate more text tokens? I know we aggressively pad in the image dimension (I think in the collate function?), but didn't think we were doing this in the text dimension

Reply:
cc @pbontrager

Reply:
This isn't a pretty solution, but essentially what's happening is during decoding your attention shapes are [1 x kv_cache_size] where 1 is the token being decoded. Inside of TransformerDecoder we take a single slice of a causal mask for our self attention masks. For cross attention, the mask is static during decoding, and the assumption being made here is the decoding tokens have the same mask as the final token of pre-fill e.g. the decoded tokens can see whatever images the previous token could.